vcd/resource_vcd_vapp_vm_tools.go

```diff
@@ -0,0 +1,1168 @@
+			if err != nil {
+				return nil, fmt.Errorf("value `%s` unit_number is not number. err: %s", unitNumber, err)
+			}
+			addParams.unitNumber = &convertedUnitNumber
+		}
+
+		diskParamsArray = append(diskParamsArray, addParams)
+	}
+	return diskParamsArray, nil
+}
+
+func getVmIndependentDisks(vm govcd.VM) []string {
+
+	var disks []string
+	// We use VirtualHardwareSection because in time of implementation we didn't have access to VmSpecSection which we used for internal disks.
+	for _, item := range vm.VM.VirtualHardwareSection.Item {
+		// disk resource type is 17
+		if item.ResourceType == 17 && item.HostResource[0].Disk != "" {
+			disks = append(disks, item.HostResource[0].Disk)
+		}
+	}
+	return disks
+}
+
+func addRemoveGuestProperties(d *schema.ResourceData, vm *govcd.VM) error {
+	if d.HasChange("guest_properties") {
+		vmProperties, err := getGuestProperties(d)
+		if err != nil {
+			return fmt.Errorf("unable to convert guest properties to data structure")
+		}
+
+		log.Printf("[TRACE] Updating VM guest properties")
+		_, err = vm.SetProductSectionList(vmProperties)
+		if err != nil {
+			return fmt.Errorf("error setting guest properties: %s", err)
+		}
+	}
+	return nil
+}
+
+// isPrimaryNicRemoved checks if new schema has a primary NIC at all
+func isPrimaryNicRemoved(d *schema.ResourceData) bool {
+	_, newNetworkRaw := d.GetChange("network")
+	newNetworks := newNetworkRaw.([]interface{})
+
+	var foundPrimaryNic bool
+	for _, newNet := range newNetworks {
+		netMap := newNet.(map[string]interface{})
+		isPrimary := netMap["is_primary"].(bool)
+		if isPrimary {
+			foundPrimaryNic = true
+			break
+		}
+	}
+
+	return !foundPrimaryNic
+}
+
+func updateVmSpecSection(vmSpecSection *types.VmSpecSection, vm *govcd.VM, description string) error {
+	// add missing values if not inherited from template, otherwise API throws error if some value is nil
+	if vmSpecSection.MemoryResourceMb.Reservation == nil {
+		vmSpecSection.MemoryResourceMb.Reservation = takeInt64Pointer(int64(0))
+	}
+	if vmSpecSection.MemoryResourceMb.Limit == nil {
+		vmSpecSection.MemoryResourceMb.Limit = takeInt64Pointer(int64(-1))
+	}
+	if vmSpecSection.MemoryResourceMb.SharesLevel == "" {
+		vmSpecSection.MemoryResourceMb.SharesLevel = "NORMAL"
+	}
+	if vmSpecSection.CpuResourceMhz.Reservation == nil {
+		vmSpecSection.CpuResourceMhz.Reservation = takeInt64Pointer(int64(0))
+	}
+	if vmSpecSection.CpuResourceMhz.Limit == nil {
+		vmSpecSection.CpuResourceMhz.Limit = takeInt64Pointer(int64(-1))
+	}
+	if vmSpecSection.CpuResourceMhz.SharesLevel == "" {
+		vmSpecSection.CpuResourceMhz.SharesLevel = "NORMAL"
+	}
+	_, err := vm.UpdateVmSpecSection(vmSpecSection, description)
+	if err != nil {
+		return fmt.Errorf("error updating Vm Spec Section: %s", err)
+	}
+	return nil
+}
+
+func getVmFromResource(d *schema.ResourceData, meta interface{}, vmType typeOfVm) (*VCDClient, *govcd.Org, *govcd.Vdc, *govcd.VApp, string, *govcd.VM, error) {
```

Comment:
Please add a comment explaining the return values

Reply:
Added. In reality it is just a shortcut to get VM and its parent structures from HCL configuration.

Reply:
I know, but I think of the ones who will manage this code beside us. :)